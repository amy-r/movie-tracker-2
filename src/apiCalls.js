import { key } from './key';

export const test = `https://api.themoviedb.org/3/discover/movie?api_key=${key}&language=en-US&sort_by=revenue.desc&include_adult=false&include_video=false&page=1`;

export const fetchParse = async url => {
  try {
    const response = await fetch(url);
    const parsed = await response.json();
    return parsed;
  } catch (error) {
    return 'fetchParse Err';
  }
};

export const movieCleaner = movies => {
  if (movies === null) {
    return undefined;
  }
  return movies.results.map(movie => {
    const {
      title,
      poster_path,
      id,
      release_date,
      vote_average,
      overview
    } = movie;
    return {
      title,
      poster_path,
      movie_id: id,
      release_date,
      vote_average,
      overview
    };
  });
};

export const logIn = async user => {
  try {
    const response = await fetch('/api/users', {
      method: 'POST',
      body: JSON.stringify(user),
      headers: { 'Content-Type': 'application/json' }
    });
    const parsed = await response.json();
    if (response.ok) {
      return parsed.data;
    }
  } catch (error) {
    return "Error in logIn";
  }
};

export const createUser = async user => {
  try {
    const response = await fetch('/api/users/new', {
      method: 'POST',
      body: JSON.stringify(user),
      headers: { 'Content-Type': 'application/json' }
    });
    const parsed = await response.json();
    if (response.ok) {
      return parsed;
    }
  } catch (error) {
    return "Error in createUser";
  }
};

export const addFavorite = async (movie) => {
  try {
     const response = await fetch('api/users/favorites/new', { 
      method: 'POST', 
      body: JSON.stringify(movie),
      headers: { 'Content-Type': 'application/json' }
    })
    const parsed = await response.json();
    if (response.ok) {
      return parsed;
    }
  } catch (error) {
    return 'Error in addFavorite';
  }   
}

export const removeFavorite = async (movie) => {
  const { movie_id, user_id } = movie;
  try {
     const response = await fetch(`api/users/${user_id}/favorites/${movie_id}`, { 
      method: 'DELETE', 
      body: JSON.stringify({ movie_id, user_id }),
      headers: { 'Content-Type': 'application/json' }
    })
    const parsed = await response.json();
    if(response.ok) {
      return parsed;
    }
  } catch (error) {
    return 'Error in removeFavorite';
  }   
}

export const fetchAllFavorites = async (userId) => {
  try {
    const response = await fetch(`api/users/${userId}/favorites`)
    const favorites = await response.json()
    return favorites.data
  } catch(error) {
      return 'Error in fetchAllFavorites';
  }
  
}