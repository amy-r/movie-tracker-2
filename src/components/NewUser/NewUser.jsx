import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { logIn } from '../../actions/actions';
import * as api from '../../apiCalls';
import PropTypes from 'prop-types';
import './NewUser.css';

export class NewUser extends Component {
  constructor() {
    super();
    this.state = {
      name: '',
      email: '',
      password: ''
    };
  }

  handleInput = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };

  handleSubmit = async event => {
    event.preventDefault();
    const response = await api.createUser(this.state);
    if (response) {
      const { email, password } = this.state; 
      const credentials = Object.assign({}, {email}, {password});
      const user = await api.logIn(credentials);
      if (user) {
        this.props.logIn(user);
        this.props.history.push('/');
      }
    } else {
      alert('Email has already been used');
      this.setState({name:'', email:'', password:''});
    }
  };

  render() {
    return (
      <div className="NewUser">
        <form onSubmit={this.handleSubmit}>
          <label htmlFor="name">Name: </label>
          <input
            type="text"
            name="name"
            onChange={this.handleInput}
            value={this.state.name}
          />
          <label htmlFor="email">Email: </label>
          <input
            type="text"
            name="email"
            onChange={this.handleInput}
            value={this.state.email}
          />
          <label htmlFor="password">Password: </label>
          <input
            type="text"
            name="password"
            value={this.state.password}
            onChange={this.handleInput}
          />
          <input type="submit" />
        </form>
      </div>
    );
  }
}

  
export const mapDispatchToProps = dispatch => ({
  logIn: user => dispatch(logIn(user))
});

NewUser.propTypes = {
  logIn: PropTypes.func,
  history: PropTypes.object
};

export default withRouter(
  connect(null, mapDispatchToProps)(NewUser)
);
