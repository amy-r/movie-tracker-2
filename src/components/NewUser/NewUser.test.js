import React from 'react';
import { NewUser , mapDispatchToProps } from './NewUser';
import { shallow, mount } from 'enzyme';
import * as api from '../../apiCalls'

describe('NEW_USER', () => {
  let wrapper;
  beforeEach( () => {
    wrapper = shallow(<NewUser />); 
  })

  describe('when rendered', () => {
    it.skip('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  })
    
  describe('handleInput', () => {
    describe('when given a key event', () => {
      let event;
      
      it('should update the state with name', () => {  
        expect(wrapper.state('name')).toEqual('')
        event = {
          target: {
            value: 'Nora',
            name: 'name'
          }
        }
        wrapper.instance().handleInput(event);
        expect(wrapper.state('name')).toEqual('Nora')
      });

      it('should update the state with email', () => {  
        expect(wrapper.state('email')).toEqual('')
        event = {
          target: {
            value: 'nora@gmail.com',
            name: 'email'
          }
        }
        wrapper.instance().handleInput(event);
        expect(wrapper.state('email')).toEqual('nora@gmail.com')
      });

      it('should update the state with password', () => {  
        expect(wrapper.state('password')).toEqual('')
        event = {
          target: {
            value: 'chicken123',
            name: 'password'
          }
        }
        wrapper.instance().handleInput(event);
        expect(wrapper.state('password')).toEqual('chicken123')
      });

    })
    
  });

  describe('handleSubmit', () => {
      let mockEvent;
      let url;
      let mockFetchParams;
      let mockData;
      let expected

    describe('when passed a state of new name, email, and password', () => {
      beforeAll( () => {

        wrapper.state = {
          name: 'Nora',
          email: 'nora@gmail.com',
          password: 'chicken123',
        };

       mockEvent = { preventDefault: jest.fn() }
      })


      it('should call createUser', async() => {
        // user = await JSON.stringify(wrapper.state);

        window.fetch = await jest.fn().mockImplementation(() => {
          return {status: "success", message: "New user created", id: 21}
        }) 
        
        wrapper.instance().handleSubmit(mockEvent);

        expect(window.fetch).toHaveBeenCalled()
      })
    })

    describe('when given an email already in the database', () => {
 
      it('should call setState with empty strings', () => {

        expected = { name: '', email: '', password: '' };

        wrapper.setState({ email: 'bar', password: 'foo' });
        wrapper.instance().newUserError();

        expect(wrapper.state()).toEqual(expected);
      })
    })

  });

  
  describe('MDTP', () => {
    let mockDispatch;
    let mapped;
    it('should call the dispatch function on MDTP', () => {
      mockDispatch = jest.fn();
      mapped = mapDispatchToProps(mockDispatch);
      mapped.logIn();
      expect(mockDispatch).toHaveBeenCalled();
    });

  })  
});