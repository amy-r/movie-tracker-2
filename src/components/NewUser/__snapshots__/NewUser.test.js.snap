// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NEW_USER when rendered should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NewUser
    history={Array []}
    logIn={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
>
        <label
                htmlFor="name"
        >
                Name: 
        </label>
        <input
                name="name"
                onChange={[Function]}
                type="text"
                value=""
        />
        <label
                htmlFor="email"
        >
                Email: 
        </label>
        <input
                name="email"
                onChange={[Function]}
                type="text"
                value=""
        />
        <label
                htmlFor="password"
        >
                Password: 
        </label>
        <input
                name="password"
                onChange={[Function]}
                type="text"
                value=""
        />
        <input
                type="submit"
        />
</form>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <label
            htmlFor="name"
>
            Name: 
</label>,
          <input
            name="name"
            onChange={[Function]}
            type="text"
            value=""
/>,
          <label
            htmlFor="email"
>
            Email: 
</label>,
          <input
            name="email"
            onChange={[Function]}
            type="text"
            value=""
/>,
          <label
            htmlFor="password"
>
            Password: 
</label>,
          <input
            name="password"
            onChange={[Function]}
            type="text"
            value=""
/>,
          <input
            type="submit"
/>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Name: ",
            "htmlFor": "name",
          },
          "ref": null,
          "rendered": "Name: ",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "name",
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Email: ",
            "htmlFor": "email",
          },
          "ref": null,
          "rendered": "Email: ",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "email",
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Password: ",
            "htmlFor": "password",
          },
          "ref": null,
          "rendered": "Password: ",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "name": "password",
            "onChange": [Function],
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "type": "submit",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
>
          <label
                    htmlFor="name"
          >
                    Name: 
          </label>
          <input
                    name="name"
                    onChange={[Function]}
                    type="text"
                    value=""
          />
          <label
                    htmlFor="email"
          >
                    Email: 
          </label>
          <input
                    name="email"
                    onChange={[Function]}
                    type="text"
                    value=""
          />
          <label
                    htmlFor="password"
          >
                    Password: 
          </label>
          <input
                    name="password"
                    onChange={[Function]}
                    type="text"
                    value=""
          />
          <input
                    type="submit"
          />
</form>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <label
              htmlFor="name"
>
              Name: 
</label>,
            <input
              name="name"
              onChange={[Function]}
              type="text"
              value=""
/>,
            <label
              htmlFor="email"
>
              Email: 
</label>,
            <input
              name="email"
              onChange={[Function]}
              type="text"
              value=""
/>,
            <label
              htmlFor="password"
>
              Password: 
</label>,
            <input
              name="password"
              onChange={[Function]}
              type="text"
              value=""
/>,
            <input
              type="submit"
/>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Name: ",
              "htmlFor": "name",
            },
            "ref": null,
            "rendered": "Name: ",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "name",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Email: ",
              "htmlFor": "email",
            },
            "ref": null,
            "rendered": "Email: ",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "email",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Password: ",
              "htmlFor": "password",
            },
            "ref": null,
            "rendered": "Password: ",
            "type": "label",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "name": "password",
              "onChange": [Function],
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "type": "submit",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
